cpp/src/parquet/column_reader_test.cc

```diff
@@ -572,5 +576,615 @@ TEST_F(TestPrimitiveReader, TestNonDictionaryEncodedPagesWithExposeEncoding) {
+  }
+
+  {
+    // Read 3 records: {20, null, 30}
+    int64_t records_read = record_reader->ReadRecords(/*num_records=*/3);
+
+    ASSERT_EQ(records_read, 3);
+    // One of these values is null. values_written() includes null values.
+    ASSERT_EQ(record_reader->values_written(), 3);
+    ASSERT_EQ(record_reader->null_count(), 1);
+    // When we read, we buffer some levels. Since there are only a few levels
+    // in our whole column, all of them are read.
+    // We had skipped 2 of the levels above. So there is only 6 left in total to
+    // read, and we read 3 of them here.
+    ASSERT_EQ(record_reader->levels_written(), 6);
+    ASSERT_EQ(record_reader->levels_position(), 3);
+
+    std::vector<int16_t> read_defs(
+        record_reader->def_levels(),
+        record_reader->def_levels() + record_reader->levels_position());
+
+    const auto read_values = reinterpret_cast<const int32_t*>(record_reader->values());
+    std::vector<int32_t> read_vals(read_values,
+                                   read_values + record_reader->values_written());
+
+    // ReadRecords for optional fields uses ReadValuesSpaced, so there is a
+    // placeholder for null.
+    ASSERT_EQ(read_vals[0], 20);
+    // read_vals[1] is a space for null.
+    ASSERT_EQ(read_vals[2], 30);
+    ASSERT_THAT(read_defs, ElementsAre(1, 0, 1));
+  }
+
+  {
+    // Skip {40, 50}.
+    int64_t records_skipped = record_reader->SkipRecords(/*num_records=*/2);
+
+    ASSERT_EQ(records_skipped, 2);
+
+    ASSERT_EQ(record_reader->values_written(), 3);
+    ASSERT_EQ(record_reader->levels_written(), 4);
+    ASSERT_EQ(record_reader->levels_position(), 3);
+
+    std::vector<int16_t> read_defs(
+        record_reader->def_levels(),
+        record_reader->def_levels() + record_reader->levels_position());
+
+    const auto read_values = reinterpret_cast<const int32_t*>(record_reader->values());
+    std::vector<int32_t> read_vals(read_values,
+                                   read_values + record_reader->values_written());
+  }
+
+  {
+    // Read to the end of the column. Read {60}
+    // This test checks that ReadAndThrowAwayValues works, since if it
+    // does not we would read the wrong values.
+    int64_t records_read = record_reader->ReadRecords(/*num_records=*/1);
+
+    ASSERT_EQ(records_read, 1);
+    ASSERT_EQ(record_reader->values_written(), 4);
+    ASSERT_EQ(record_reader->null_count(), 1);
+    ASSERT_EQ(record_reader->levels_written(), 4);
+    ASSERT_EQ(record_reader->levels_position(), 4);
+
+    std::vector<int16_t> read_defs(
+        record_reader->def_levels(),
+        record_reader->def_levels() + record_reader->levels_position());
+
+    const auto read_values = reinterpret_cast<const int32_t*>(record_reader->values());
+    std::vector<int32_t> read_vals(read_values,
+                                   read_values + record_reader->values_written());
+
+    ASSERT_EQ(read_vals[0], 20);
+    // read_vals[1] is a space for null.
```

Comment:
Does it have an expected value anyway? I would expect the implementation to zero it out.